pulsar/internal/connection.go

```diff
@@ -309,7 +310,13 @@ func (c *connection) WriteData(data []byte) {
 
 func (c *connection) internalWriteData(data []byte) {
 	c.log.Debug("Write data: ", len(data))
-	if _, err := c.cnx.Write(data); err != nil {
+	err := c.cnx.SetWriteDeadline(time.Now().Add(writeDeadline))
```

Comment:
Shouldn't this be set just once per each connection? doing it here would set it each time we write on the connection. 

Also, since this will affect the publish timeouts, it would be better to keep it configurable (as a client option). 

Finally, what is the exact meaning of the write deadline? Will it fail the write op if the connection is not writable for the specified amount of time? I think we are already covering this scenario with the keep-alive mechanism that detects either a clogged or stale TCP connection.

Reply:
Should be set before each Write.

Connections produced via the connection pool and clients are not aware of the underlying configuration. Exposing such low-level config is redundant, we can increase the timeout though.

If a connection got stuck on Write, it will not reach the ping ticker message. Also, it will not release the Flush WaitGroup (impl_partition_producer.go: 438).

@merlimat  WDYT?